pkg/volume/csi/csi_test.go

```diff
@@ -0,0 +1,339 @@
+				return volume.NewSpecFromVolume(makeTestVol(specName, driver))
+			},
+			podFunc: func() *api.Pod {
+				podUID := types.UID(fmt.Sprintf("%08X", rand.Uint64()))
+				return &api.Pod{ObjectMeta: meta.ObjectMeta{UID: podUID, Namespace: testns}}
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			tmpDir, err := utiltesting.MkTmpdir("csi-test")
+			if err != nil {
+				t.Fatalf("can't create temp dir: %v", err)
+			}
+			defer os.RemoveAll(tmpDir)
+			client := fakeclient.NewSimpleClientset()
+			fakeWatcher := watch.NewRaceFreeFake()
+
+			factory := informers.NewSharedInformerFactory(client, csiResyncPeriod)
+			go factory.Start(wait.NeverStop)
+
+			host := volumetest.NewFakeVolumeHostWithCSINodeName(
+				tmpDir,
+				client,
+				nil,
+				"csi-node",
+				factory.Storage().V1beta1().CSIDrivers().Lister(),
+			)
+
+			plugMgr := &volume.VolumePluginMgr{}
+			plugMgr.InitPlugins(ProbeVolumePlugins(), nil /* prober */, host)
+			csiClient := setupClient(t, true)
+
+			volSpec := test.specFunc(test.specName, test.driver, test.volName)
+			pod := test.podFunc()
+			attachName := getAttachmentName(test.volName, test.driver, string(host.GetNodeName()))
+			t.Log("csiTest.VolumeAll starting...")
+
+			// *************** Attach/Mount volume resources ****************//
+			// attach volume
+			t.Log("csiTest.VolumeAll Attaching volume...")
+			attachPlug, err := plugMgr.FindAttachablePluginBySpec(volSpec)
+			if err != nil {
+				t.Fatalf("csiTest.VolumeAll PluginManager.FindAttachablePluginBySpec failed: %v", err)
+			}
+
+			var devicePath string
+			if attachPlug != nil {
+				t.Log("csiTest.VolumeAll attacher.Attach starting")
+
+				var volAttacher volume.Attacher
+
+				volAttacher, err := attachPlug.NewAttacher()
+				if err != nil {
+					t.Fatal("csiTest.VolumeAll failed to create new attacher: ", err)
+				}
+
+				go func(spec *volume.Spec, nodeName types.NodeName) {
+					attachID, err := volAttacher.Attach(spec, nodeName)
+					if err != nil {
+						t.Fatalf("csiTest.VolumeAll attacher.Attach failed: %s", err)
+					}
+					t.Logf("csiTest.VolumeAll got attachID %s", attachID)
+
+				}(volSpec, host.GetNodeName())
+
+				markVolumeAttached(t, host.GetKubeClient(), fakeWatcher, attachName, storage.VolumeAttachmentStatus{Attached: true})
+
+				devicePath, err = volAttacher.WaitForAttach(volSpec, "", pod, 500*time.Millisecond)
+				if err != nil {
+					t.Fatal("csiTest.VolumeAll attacher.WaitForAttach failed:", err)
+				}
+
+				if devicePath != attachName {
+					t.Fatalf("csiTest.VolumeAll attacher.WaitForAttac got unexpected value %s", devicePath)
+				}
+
+				t.Log("csiTest.VolumeAll attacher.WaitForAttach succeeded OK, attachment ID:", devicePath)
+
+			} else {
+				t.Log("csiTest.VolumeAll volume attacher not found, skipping attachment")
+			}
+
+			// Mount Device
+			t.Log("csiTest.VolumeAll Mouting device...")
+			devicePlug, err := plugMgr.FindDeviceMountablePluginBySpec(volSpec)
```

Comment:
you should check that inline volume does not get devicePlug here.

Reply:
done.